hathor/feature_activation/feature_service.py

```diff
@@ -54,6 +82,9 @@ def get_state(self, *, block: Block, feature: Feature) -> FeatureState:
         previous_boundary_block = self._get_ancestor_at_height(block=block, height=previous_boundary_height)
         previous_boundary_state = self.get_state(block=previous_boundary_block, feature=feature)
 
+        # We cache _and save_ the state of the previous boundary block that we just got.
+        previous_boundary_block.set_feature_state(feature=feature, state=previous_boundary_state, save=True)
```

Comment:
Why are we changing the metadata in a `get_state()` method? I feel that this method is not supposed to change anything.

Reply:
This is the caching mechanism created to prevent repeating long-running calculations of feature states for blocks. Without caching, each time we want to cache for a block's state, we would have to calculate the state of all boundary blocks down to the genesis. This cache is stored in metadata.

Saving this in the `get_state()` method is actually not new,  it was firstly implemented in https://github.com/HathorNetwork/hathor-core/pull/645. Once a block is received and its state is calculated, it never changes. It's fixed forever, as it only depends on its ancestors. Therefore, this `get_state()` works like a lazy calculation: as soon as the state is queried for the first time, it's calculated and persisted in metadata. Then, any subsequent calls simply return the saved state.

The change here is that before, we were saving the state of the new, just queried block. This was the straightforward solution. However, now the `get_state()` method is used indirectly during the block verification process. That occurs before the block's `metadata.validation` is marked, so I cannot save the metadata at this point. Now, to workaround this, I delayed the caching. Instead of saving the current block's metadata in the storage, I only save it in memory. But I also save the previous boundary block's state in the storage (which is the line you commented).

Reply:
Got it. Thanks for the detailed explanation. It makes sense to store cache information. Should we add a note to this metadata saying that it's a cache and can be safely cleared up?

Reply:
Done in https://github.com/HathorNetwork/hathor-core/pull/785/commits/c07032019b78c1ce3cbcaa067d1d65748c337353